src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -0,0 +1,539 @@
+        {
+          slotDur = it->second - it->first;
+
+          if (slotDur > allocDuration)
+            {
+
+              endAlloc = AllocateSingleContiguousBlock (info.GetAllocationID (), info.GetAllocationType (), info.IsPseudoStatic (),
+                                                        sourceAid, info.GetDestinationAid (), it->first, allocDuration);
+              m_remainingDtiTime -= (allocDuration + m_guardTime);
+
+              // The following function modifies m_availableSlots, on which this
+              // loop is iterating. Commonly, it is a bad practice but, as we break 
+              // the loop if we enter this condition, the damage is avoided.
+              UpdateAvailableSlots (it->first, endAlloc);
+              status.SetSuccess ();
+              break;
+            }
+          else
+            {
+              status.SetFailure ();
+            }
+        }
+
+    }
+
+  return status;
+}
+
+uint8_t
+PeriodicDmgWifiScheduler::GetAvailableBlocks (uint32_t allocDuration, uint32_t spInterval)
+{
+  NS_LOG_FUNCTION (this);
+
+  auto it = m_availableSlots.begin ();
+  uint32_t startAlloc = it->first;
+  uint32_t slotDuration = 0;
+  uint8_t blocks = 0;
+
+  while (it != m_availableSlots.end ())
+    {
+      if (startAlloc < it->first)
+        {
+          // Periodicity has been broken, we stop the algorithm.
+          break;
+        }
+
+      slotDuration = it->second - startAlloc;
+
+      if (allocDuration + m_guardTime > slotDuration)
+        {
+          if (blocks == 0)
+            {
+              // go on until eventually find the first available slot that fits this SP
+              // note that this also cover the condition where no slot satisfies the requirement
+              ++it;
+              startAlloc = it->first;
+              continue;
+            }
+          else
+            {
+              // if one or more periodic SPs already allocated, now the periodicity
+              // is broken and the algorithm stops
+              break;
+            }
+        }
+
+      blocks++;
+      startAlloc += spInterval;
+
+      // immediately check the next periodic allocation, with respect to the
+      // current available slot that we are using
+
+      if (startAlloc == it->second)
+        {
+          // if the next SP should start at the end of the current slot, 
+          // it cannot be added and the algorithm stops
+          break;
+        }
+      else if (startAlloc < it->second)
+        {
+          // if the the next SP should start in the current slot, immediately check if there is
+          // enough free time to place it, otherwise the algorithm stops
+          if (startAlloc + allocDuration + m_guardTime > it->second)
+            {
+              break;
+            }
+        }
+      else
+        {
+          // pass to the next slot only if startAlloc goes beyond the end of the current one
+          ++it;
+        }
```

Comment:
There's no need to check these things, you will do it during the next loop. You are basically copying the code from the previous if statement in a slightly different way. If later on you find a bug you'll have to fix both conditional statements: chances are that you (or anybody) will forget about it and replace a bug with another one, probably more tricky to find

```suggestion
      // if next allocation period exceeds the current available slot's boundaries
      // proceed to the next one
      if (startAlloc > it->second)
        {
          ++it;
        }
      // else: the next allocation period is still (partially) in the current
      // available slot
```